src/System.Threading/tests/AsyncLocalTests.cs

```diff
@@ -419,5 +423,162 @@ private static Task Run(Func<Task> func)
                 }
             }
         }
+
+        [Theory]
+        [MemberData(nameof(GetCount))]
+        public static async Task AsyncLocalsUnwind(int count)
+        {
+            AsyncLocal<object>[] locals = new AsyncLocal<object>[count];
+
+            int[] setManually = new int[count];
+            int[] unsetAutomatically = new int[count];
+            int[] setAutomatically = new int[count];
+            ExecutionContext[] contexts = new ExecutionContext[count];
+
+            await AsyncRecursive(count - 1, locals, contexts, args =>
+            {
+                if (!args.ThreadContextChanged)
+                {
+                    if (args.PreviousValue != null)
+                    {
+                        Assert.True(false, $"Trying to set {args.CurrentValue} when {args.PreviousValue} value already set");
+                    }
+
+                    Assert.IsType<int>(args.CurrentValue);
+
+                    setManually[(int)args.CurrentValue]++;
+                }
+                else
+                {
+                    if (args.PreviousValue == null && args.CurrentValue == null)
+                    {
+                        Assert.True(false, "CurrentValue and PreviousValue are both null");
+                    }
+
+                    if (args.PreviousValue != null && args.CurrentValue != null)
+                    {
+                        Assert.True(false, "CurrentValue and PreviousValue are both set");
+                    }
+
+                    if (args.CurrentValue != null)
+                    {
+                        Assert.IsType<int>(args.CurrentValue);
+                        setAutomatically[(int)args.CurrentValue]++;
+                    }
+                    else
+                    {
+                        Assert.IsType<int>(args.PreviousValue);
+                        unsetAutomatically[(int)args.PreviousValue]++;
+                    }
+
+                }
+            });
+
+            for (int i = 0; i < locals.Length; i++)
+            {
+                Assert.Equal(1, setManually[i]);
+                Assert.Equal(1, unsetAutomatically[i]);
+                Assert.Equal(0, setAutomatically[i]);
+            }
+
+            Array.Clear(setManually, 0, count);
+            Array.Clear(unsetAutomatically, 0, count);
+            Array.Clear(setAutomatically, 0, count);
+
+            ExecutionContext Default = ExecutionContext.Capture();
+
+            TestExecutionContextRun();
+
+            ExecutionContext.SuppressFlow();
+            try
+            {
+                TestExecutionContextRun();
+            }
+            finally
+            {
+                ExecutionContext.RestoreFlow();
+            }
```

Comment:
There's a lot happening in this one test; can it be split apart?

Reply:
Refactored into locals and added comments; would be a lot of setup repetition if I split the test; or lots of params for functions.

Better now though; or should I split?